packages/features/form-builder/FormBuilder.tsx

```diff
@@ -0,0 +1,737 @@
+          placeholder={field.placeholder}
+          label={field.label}
+          readOnly={readOnly}
+          name={field.name}
+          value={value as string}
+          setValue={setValue as (arg: typeof value) => void}
+        />
+      </WithLabel>
+    );
+  }
+
+  if (componentConfig.propsType === "boolean") {
+    return (
+      <WithLabel field={field} readOnly={readOnly}>
+        <componentConfig.factory
+          label={field.label}
+          readOnly={readOnly}
+          value={value as boolean}
+          setValue={setValue as (arg: typeof value) => void}
+          placeholder={field.placeholder}
+        />
+      </WithLabel>
+    );
+  }
+
+  if (componentConfig.propsType === "textList") {
+    return (
+      <WithLabel field={field} readOnly={readOnly}>
+        <componentConfig.factory
+          placeholder={field.placeholder}
+          label={field.label}
+          readOnly={readOnly}
+          value={value as string[]}
+          setValue={setValue as (arg: typeof value) => void}
+        />
+      </WithLabel>
+    );
+  }
+
+  if (componentConfig.propsType === "select") {
+    if (!field.options) {
+      throw new Error("Field options is not defined");
+    }
+
+    return (
+      <WithLabel field={field} readOnly={readOnly}>
+        <componentConfig.factory
+          readOnly={readOnly}
+          value={value as string}
+          placeholder={field.placeholder}
+          setValue={setValue as (arg: typeof value) => void}
+          options={field.options.map((o) => ({ ...o, title: o.label }))}
+        />
+      </WithLabel>
+    );
+  }
+
+  if (componentConfig.propsType === "multiselect") {
+    if (!field.options) {
+      throw new Error("Field options is not defined");
+    }
+    return (
+      <WithLabel field={field} readOnly={readOnly}>
+        <componentConfig.factory
+          placeholder={field.placeholder}
+          readOnly={readOnly}
+          value={value as string[]}
+          setValue={setValue as (arg: typeof value) => void}
+          options={field.options.map((o) => ({ ...o, title: o.label }))}
+        />
+      </WithLabel>
+    );
+  }
+
+  if (componentConfig.propsType === "objectiveWithInput") {
+    if (!field.options) {
+      throw new Error("Field options is not defined");
+    }
+    if (!field.optionsInputs) {
+      throw new Error("Field optionsInputs is not defined");
+    }
+    return field.options.length ? (
+      <WithLabel field={field} readOnly={readOnly}>
+        <componentConfig.factory
+          placeholder={field.placeholder}
+          readOnly={readOnly}
+          name={field.name}
+          value={value as { value: string; optionValue: string }}
+          setValue={setValue as (arg: typeof value) => void}
+          optionsInputs={field.optionsInputs}
+          options={field.options}
+        />
+      </WithLabel>
+    ) : null;
+  }
+
+  throw new Error(`Field ${field.name} does not have a valid propsType`);
```

Comment:
I'm pretty sure we can improve a lot in terms of legibility and typing here.

Reply:
Yeah I think so too. But couldn't get type narrowing to work and had to resort to type assertion. Will followup on it.